File: tools/java/crcl4java/examples/C++/crclServer.cpp

#include "crclj.h"
#include <iostream>

using namespace crclj;
using namespace crclj::crcl::base;
using namespace crclj::crcl::utils;
using namespace crclj::java::math;
using namespace crclj::java::util;
using namespace crclj::java::lang;

using namespace std;

static bool quit=false;


int main(int argc, const char **argv) {

    try {
        
        
        // Create a status message for sending to clients.
        CRCLStatusType status;
        CommandStatusType cmdStatus;
        cmdStatus.setCommandID(1);
        cmdStatus.setStatusID(1);
        status.setCommandStatus(cmdStatus);
        PoseStatusType poseStatus;

        // Create as Server Socket object bound to the default port.

        CRCLServerSocket serverSocket(CRCLSocket::getDEFAULT_PORT());

        // Enable queuing of events so we don't need a callback.
        serverSocket.setQueueEvents(true);

        
        // Start a background thread to listen and accept connections.
        serverSocket.start();
        cout << " C++ CRCL Server background thread started. " << endl;
        
        int requestCount = 1;

        // Process events until someone interrupts us.
        while (!quit) {

            // Wait for the next event.
            CRCLServerSocketEvent e = serverSocket.waitForEvent();

            // Get an object representing a connection back to the 
            // client responsible for this event.
            CRCLSocket crclSocket = e.getSource();

            // Check if this event has a Command instance attached.
            CRCLCommandInstanceType instance = e.getInstance();
            if (NULL != instance.jthis) {

                //Get the command from the instance.
                CRCLCommandType cmd = instance.getCRCLCommand();
                if (GetStatusType::instanceof(cmd)) {

                    // In this simple simulated example all commands are
                    // DONE after 3 get status requests. 
                    // In a more realistic system, some commands like move a long
                    // distance would take a long time and some commands that
                    // just change a setting will immediately be done.
                    requestCount++;
                    if (requestCount > 3) {
                        cmdStatus.setCommandState(CommandStateEnumType::getCRCL_DONE());
                    }
                    cmdStatus.setStatusID(requestCount);
                    crclSocket.writeStatus(status, true);
                } else {
                    requestCount = 1;
                    
                    // Set the CommandID so client can know which command we are
                    // working on.
                    cmdStatus.setCommandID(cmd.getCommandID());
                    cmdStatus.setCommandState(CommandStateEnumType::getCRCL_WORKING());
                    if (MoveToType::instanceof(cmd)) {

                        // Simulates a system where the goal endPosition is immediately
                        // reached.
                        poseStatus.setPose(MoveToType::cast(cmd).getEndPosition());
                        status.setPoseStatus(poseStatus);
                    } else if(MessageType::instanceof(cmd)) {
                        MessageType messageCmd = MessageType::cast(cmd);
                        jstring msgJString = messageCmd.getMessage();
                        jboolean isCopy = false;
                        const char *msgCString = GetStringUTFChars(msgJString,&isCopy);
                        cout << "message=" << msgCString <<endl;
                        ReleaseStringUTFChars(msgJString,msgCString);
                    }
                }
            }

            // Events could also have an exception attached that could be
            // handled here.
            Exception exception = e.getException();
            if (NULL != exception.jthis) {
                PrintObject("exception",exception);
            }
        }
        cout << " Closing serverSocket ... " << endl;
        // Close all connected sockets and interrupt() & join() any background threads
        serverSocket.close();
        
        cout << " End of C++ CRCL Server main() reached. " << endl;
    } catch (jthrowable t) {
        PrintJThrowable("Exception Thrown : ", t);
    }
}

File: tools/java/crcl4java/examples/java/CRCLServer.java
/* 
 * This is public domain software, however it is preferred
 * that the following disclaimers be attached.
 * 
 * Software Copywrite/Warranty Disclaimer
 * 
 * This software was developed at the National Institute of Standards and
 * Technology by employees of the Federal Government in the course of their
 * official duties. Pursuant to title 17 Section 105 of the United States
 * Code this software is not subject to copyright protection and is in the
 * public domain. This software is experimental.
 * NIST assumes no responsibility whatsoever for its use by other
 * parties, and makes no guarantees, expressed or implied, about its
 * quality, reliability, or any other characteristic. We would appreciate
 * acknowledgment if the software is used. This software can be
 * redistributed and/or modified freely provided that any derivative works
 * bear some notice that they are derived from it, and any modified
 * versions bear some notice that they have been modified.
 * 
 */
import crcl.base.*;
import crcl.utils.CRCLSocket;
import java.io.IOException;
import java.math.BigInteger;
import java.net.ServerSocket;
import java.net.Socket;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 *
 * @author Will Shackleford{@literal <william.shackleford@nist.gov> }
 */
public class CRCLServer {

    public static void main(String[] args) throws IOException {

        final CRCLStatusType status = new CRCLStatusType();
        final CommandStatusType cmdStatus = new CommandStatusType();
        cmdStatus.setCommandID(1);
        cmdStatus.setStatusID(1);
        status.setCommandStatus(cmdStatus);
        final PoseStatusType poseStatus = new PoseStatusType();
        
        final ServerSocket serverSocket = new ServerSocket(CRCLSocket.DEFAULT_PORT);
        final ExecutorService execService = Executors.newCachedThreadPool();
        execService.submit(new Runnable() {

            @Override
            public void run() {
                while (true) {
                    try {
                        final Socket socket = serverSocket.accept();
                        execService.submit(new Runnable() {
                            @Override
                            public void run() {
                                try {
                                    CRCLSocket crclSocket = new CRCLSocket(socket);
                                    CRCLCommandInstanceType instance = null;
                                    int requestCount = 1;
                                    while (null != (instance = crclSocket.readCommand(true))) {
                                        CRCLCommandType cmd = instance.getCRCLCommand();
                                        if(cmd instanceof GetStatusType) {
                                            requestCount++;
                                            if(requestCount > 3) {
                                                cmdStatus.setCommandState(CommandStateEnumType.CRCL_DONE);
                                            }
                                            cmdStatus.setStatusID(requestCount);
                                            crclSocket.writeStatus(status, true);
                                        } else {
                                            cmdStatus.setCommandID(cmd.getCommandID());
                                            cmdStatus.setCommandState(CommandStateEnumType.CRCL_WORKING);
                                            if(cmd instanceof MoveToType) {
                                                poseStatus.setPose(((MoveToType)cmd).getEndPosition());
                                                status.setPoseStatus(poseStatus);
                                            }
                                        }
                                    }
                                } catch (Exception ex) {
                                    Logger.getLogger(CRCLServer.class.getName()).log(Level.SEVERE, null, ex);
                                }
                            }
                        });
                    } catch (Exception ex) {
                        Logger.getLogger(CRCLServer.class.getName()).log(Level.SEVERE, null, ex);
                    }
                }

            }
        });
    }
}
